### PR TITLE
CR-1075730 softkernel crash when gdb attached

### PR DESCRIPTION
**Pull request by xdavidz**
softkernel to support xclSKOpenContext and xslSKCloseContext

Still testing, using pipeline to find basic issues.

### CODE REVIEW

**Review by mamin506 — APPROVED** [review]

File: src/runtime_src/core/common/drv/include/kds_core.h [hunk 1]
```diff
@@ -65,6 +65,7 @@ struct kds_ctx_info {
  * @cu_bitmap: bitmap of opening CU
  * @waitq: Wait queue for poll client
  * @event: Events to notify user client
+ * @sk_cu_id: soft kernel cu index, default is -1
```
Comment (larry9523) [inline src/runtime_src/core/common/drv/include/kds_core.h:68]: how can u32 type be default -1?

File: src/runtime_src/core/edge/drm/zocl/include/sched_exec.h [hunk 2]
```diff
@@ -76,6 +76,7 @@ struct sched_ops;
  * @abort: Flag to indicate that this context has detached from user space
  * @shcus: CUs shared reserved by this context
  * @excus: CUs exclusively reserved by this context
+ * @sk_cu_id: soft kernel cu index, default is -1
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/include/sched_exec.h:79]: see above

File: src/runtime_src/core/edge/drm/zocl/zocl_drv.c [hunk 3]
```diff
@@ -604,6 +605,10 @@ static void zocl_client_release(struct drm_device *dev, struct drm_file *filp)
 	if (!client)
 		return;
 
+	if (client->sk_cu_id != INVALID_CU_ID) {
+		zocl_fini_soft_kernel_cu(zdev, client->sk_cu_id);
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_drv.c:609]: you probably don't need {} for a single line after "if"

File: src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c [hunk 4]
```diff
@@ -66,15 +67,23 @@ zocl_ctx_ioctl(struct drm_device *ddev, void *data, struct drm_file *filp)
 		return zocl_context_ioctl(zdev, data, filp);
 	}
 
-	if (args->op == ZOCL_CTX_OP_OPEN_GCU_FD) {
+	switch(args->op) {
+	case ZOCL_CTX_OP_SK_OPEN:
+		((struct sched_client_ctx *)filp->driver_priv)->sk_cu_id = args->cu_index;
+		break;
+	case ZOCL_CTX_OP_SK_CLOSE:
+		((struct sched_client_ctx *)filp->driver_priv)->sk_cu_id = INVALID_CU_ID;
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_ioctl.c:75]: Do we need mutex protection here? Or are we relying on softkernel framework to enforce it? If latter, which is not very reliable, we should at least put some comments here

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c [hunk 5]
```diff
@@ -216,3 +218,34 @@ zocl_fini_soft_kernel(struct drm_device *drm)
...
+	if (cu_idx == INVALID_CU_ID) {
+		DRM_INFO("Invalid cu_id %u.\n", cu_idx); 
+		return -EINVAL;
+	}
+
+	if (!sk) {
+		DRM_INFO("No soft kernel.\n"); 
+		return -ENODEV;
+	}
+
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/zocl_sk.c:238]: Operation after this should be protected by mutex

**Review by larry9523 — CHANGES_REQUESTED** [review]
Minor changes requested

File: src/runtime_src/core/edge/drm/zocl/zocl_kds.c [hunk 6]
```diff
@@ -194,6 +195,12 @@ int zocl_context_ioctl(struct drm_zocl_dev *zdev, void *data,
 	case ZOCL_CTX_OP_FREE_CTX:
 		ret = zocl_del_context(zdev, client, args);
 		break;
+	case ZOCL_CTX_OP_SK_OPEN:
+		client->sk_cu_id = args->cu_index;
+		break;
+	case ZOCL_CTX_OP_SK_CLOSE:
+		client->sk_cu_id = INVALID_CU_ID;
+		break;
```
Comment (houlz0507) [inline src/runtime_src/core/edge/drm/zocl/zocl_kds.c:203]: Why do we introduce OPEN/CLOSE here but not use ALLOC_CTX and FREE_CTX?

File: src/runtime_src/core/edge/drm/zocl/zocl_sk.c [hunk 7]
```diff
@@ -216,3 +218,34 @@ zocl_fini_soft_kernel(struct drm_device *drm)
...
+	if (cu_idx >= sk->sk_ncus)
+		return -EINVAL;
+
+	scu = sk->sk_cu[cu_idx];
+	if (!scu)
+		return -ENODEV;
+
+	ZOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(scu->gem_obj);
+	kfree(scu);
+	scu = NULL;
```
Comment (houlz0507) [inline src/runtime_src/core/edge/drm/zocl/zocl_sk.c:248]: you need to set sk->sk_cu[cu_idx] to NULL here.

**Comment by xdavidz** [comment]
Not an appropriate fix, based on testing, skd will open/close fd multiple times without close the softkernel. The softkernel creataion is done by another ioctl. Thus, handling this in open/close is over kill.

**Pull request closed without merging**